src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```
@@ -336,51 +261,183 @@ void parseCvrFile(List<CastVoteRecord> castVoteRecords)
+          if (candidateIds == null || candidateIds.size() == 0) {
+            Logger.severe("CandidateSelection \"%s\" has no CandidateIds!", contestSelectionId);
+            throw new CvrParseException();
+          }
+          if (candidateIds.size() > 1) {
+            Logger.warning("CandidateSelection \"%s\" has multiple CandidateIds.  "
+                + "Only the first one will be processed.", contestSelectionId);
+          }
+          String candidateObjectId = (String) candidateIds.get(0);
+          HashMap Candidate = (HashMap) candidates.get(candidateObjectId);
+          candidateId = (String) Candidate.get("Name");
+          if (candidateId.equals(config.getOvervoteLabel())) {
+            candidateId = Tabulator.EXPLICIT_OVERVOTE_LABEL;
+          }
+          if (!this.config.getCandidateCodeList().contains(candidateId)) {
+            Logger.severe("Unrecognized candidate found in CVR: %s", candidateId);
+            unrecognizedCandidateCounts.merge(candidateId, 1, Integer::sum);
+          }
+        }
+
+        // parse rankings
+        // rank may appear on the CVRContestSelection OR the SelectionPosition
+        // this is an ambiguity in the nist spec
+        if (cvrContestSelection.containsKey("Rank")) {
+          Integer rank = (Integer) (cvrContestSelection.get("Rank"));
+          rankings.add(new Pair<>(rank, candidateId));
+        } else {
+          // extract all the SelectionPositions (ranks) which this selection has been assigned
+          ArrayList selectionPositions = (ArrayList) cvrContestSelection.get("SelectionPosition");
+          for (Object selectionPositionObject : selectionPositions) {
+            HashMap selectionPosition = (HashMap) selectionPositionObject;
+            // WriteIn can be linked at the selection position level
+            if (selectionPosition.containsKey("CVRWriteIn")) {
+              candidateId = this.config.getUndeclaredWriteInLabel();
```

what if `getUndeclaredWriteInLabel()` returns null?

---

@tarheel what do you think we should do?  Require that CDF tabulations have an UWI label?

---

I would just throw an error here if the UWI label is null.